README
fuse filesystem interface to the Arexx BS-500 data logger.

Copyright © 2013 John Hedges <john@drystone.co.uk>

Depends on libusb (>=1.0) and libfuse (>=2.9)

To run
$ arexxfs <mountpoint>

A file is created in the mountpoint for each of the tempereature/humidity
sensors. These files contain the raw 4-digit hex reading from the sensors.

To convert the hex values to useful values see:

http://www.algorithm-forge.com/techblog/2010/01/linux-usb-driver-for-the-arexx-tl-500-%E2%80%93-part-ii/

    "Further investigation shows us that we can calculate the measurements from
    RH sensors by (0.6 + Rawdata * 0.03328) RH%, from temperature sensors
    TSN-TH70E by (-39.58 + Rawdata * 0.01) °C and from temperature sensors
    TL-3TSN by (Rawdata * 0.0078) °C. These coefficients are based on the the
    following linear regressions. With more data they will slightly change, but
    given the accuracy and that you need to calibrate your sensors either way,
    it’s a sufficient translation of uncalibrated sensor data."


With thanks to Kornelius Rohmeyer for his work in decoding the tlx00 USB protocol:
http://www.algorithm-forge.com/techblog/2009/12/using-libusb-to-write-a-linux-usb-driver-for-the-arexx-tl-500-part-i/

See License for licensing.

See INSTALL for installation instructions.

Bug reports to John Hedges <john@drystone.co.uk>


ChangeLog
0.2.1 Drops the init script

0.2.0 No longer a kernel module

    * uses libusb-1.0 and fuse
    * fuse files have 4 digit hex notation of raw device reading

0.1.0 Builds with linux 3.0

    * builds with 3.0 as well as 2.6 kernel
    * moves sysfx directory from /sys/kernel/tlx00 to /sys/tlx00
    * disables debugging messages


device.c
/*
This file is part of arexxfs, a fuse interface to the arexx data loggers

Copyright © 2013 John Hedges <john@drystone.co.uk>

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

#include <stdio.h>
#include <sys/types.h>
#include <unistd.h>
#include <libusb.h>
#include <string.h>
#include <stdlib.h>
#include <syslog.h>

#include <pthread.h>

#include "device.h"

#define VENDOR  0x0451
#define PRODUCT 0x3211
#define ENDPOINT_DOWN 0x01
#define ENDPOINT_UP   0x81
#define IOLEN   64

time_t tlx_mtime, tlx_ctime;

static tlx_reading * _root_reading = NULL;
static int tlx_running = 0;

static void _start_tlx_thread();

void tlx_init() {
    _start_tlx_thread();
}

tlx_reading * tlx_get_root() {
    _start_tlx_thread();
    return _root_reading;
}

tlx_reading * tlx_get_reading(const char * sid)
{
    // make sure tlx device is still communicating
    _start_tlx_thread();

    char * e;
    long int id = strtol(sid, &e, 10);
    if (*e != '\0')
        return NULL;    // strtol didn't convert whole sid

    tlx_reading * p = _root_reading;
    while (p) {
        if ((long int)p->id == id)
            return p;
        p = p->next;
    }

    return NULL;
}

static int _record_reading(unsigned int id, unsigned int raw)
{
    tlx_reading * p = _root_reading;

    while (p) {
        if (p->id == id) {
            p->raw = raw;
            p->mtime = time(NULL);
            return 0;
        }
        p = p->next;
    }

    p = (tlx_reading *)malloc(sizeof(tlx_reading));
    if (!p) {
        syslog(LOG_DAEMON | LOG_ERR, "malloc failed (%m)");
        return -1;
    }
    p->id = id;
    p->raw = raw;
    p->ctime = p->mtime = tlx_mtime = time(NULL);
    p->next = _root_reading;
    _root_reading = p;
    return 0;
}

static void _tlx_loop(libusb_device_handle * handle)
{
    int transferred;
    unsigned char iobuf[IOLEN];

    iobuf[0] = 4; // initialise
    while (1) {
        if (libusb_bulk_transfer(handle, ENDPOINT_DOWN, iobuf, IOLEN, &transferred, 0) < 0) {
            syslog(LOG_DAEMON | LOG_CRIT, "libusb_bulk_transfer down failed (%m)");
            break;
        }
        if (libusb_bulk_transfer(handle, ENDPOINT_UP, iobuf, IOLEN, &transferred, 0) < 0) {
            syslog(LOG_DAEMON | LOG_CRIT, "libusb_bulk_transfer up failed (%m)");
            break;
        }
        if (transferred == IOLEN && (iobuf[0] || iobuf[1])) {
            if (_record_reading((iobuf[3]<<8) + iobuf[2], (iobuf[4]<<8) + iobuf[5]) < 0)
                break;
        } else sleep(1);
        iobuf[0] = 3; // read
    }
}

void * _tlx_thread(void * arg) {
    libusb_device **devs, **pdev;
    libusb_device_handle *handle;
    (void) arg;

    // set mtime and ctime for directory
    tlx_ctime = tlx_mtime = time(NULL);
    
    if (libusb_init(NULL) < 0) {
        syslog(LOG_DAEMON | LOG_CRIT, "libusb_init failed (%m)");
        goto exit2;
    }

    if (libusb_get_device_list(NULL, &devs) < 0) {
        syslog(LOG_DAEMON | LOG_CRIT, "libusb_get_device_list failed (%m)");
        goto exit1;
    }

    for(pdev = devs; *pdev != NULL; ++pdev) {
        struct libusb_device_descriptor udd;
        memset(&udd, 0, sizeof(udd));
        libusb_get_device_descriptor(*pdev, &udd);
        if (udd.idVendor == VENDOR && udd.idProduct == PRODUCT)
            break;
    }

    if (!*pdev) {
        syslog(LOG_DAEMON | LOG_CRIT, "arexx data logging device not found");
        goto exit;
    }

    if (libusb_open(*pdev, &handle) < 0) {
        syslog(LOG_DAEMON | LOG_CRIT, "libusb_open failed (%m)");
        goto exit;
    }

    syslog(LOG_DAEMON | LOG_INFO, "arexx data logging device active");

    if (libusb_kernel_driver_active(handle, 0) == 1)
        libusb_detach_kernel_driver(handle, 0);
    libusb_claim_interface(handle, 0);
    _tlx_loop(handle);
    libusb_close(handle);

exit:
    libusb_free_device_list(devs, 1);
exit1:
    libusb_exit(NULL);
exit2:
    tlx_running = 0;
    return NULL;
}

static void _start_tlx_thread()
{
    static pthread_mutex_t _start_tlx_mutex = PTHREAD_MUTEX_INITIALIZER;

    if (!tlx_running) {
        pthread_mutex_lock(&_start_tlx_mutex);
        if (!tlx_running) {
            tlx_running = 1;
            pthread_t th;
            pthread_create(&th, NULL, _tlx_thread, NULL);
        }
        pthread_mutex_unlock(&_start_tlx_mutex);
    }
}



device.h
/*
This file is part of arexxfs, a fuse interface to the arexx data loggers

Copyright © 2013 John Hedges <john@drystone.co.uk>

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

typedef struct _reading tlx_reading;

struct _reading {
    unsigned int id;
    int raw;
    time_t ctime;
    time_t mtime;
    tlx_reading * next;
};

extern time_t tlx_ctime, tlx_mtime;

void tlx_init();
tlx_reading * tlx_get_root();
tlx_reading * tlx_get_reading(const char * id);


arexxfs.c
/*
This file is part of arexxfs, a fuse interface to the arexx data loggers

Copyright © 2013 John Hedges <john@drystone.co.uk>

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <stdlib.h>

#include <fuse.h>

#include "device.h"

static int _getattr(const char *path, struct stat *stbuf)
{
    memset(stbuf, 0, sizeof(struct stat));

    if (strcmp(path, "/") == 0) {
        stbuf->st_mode = S_IFDIR | 0555;
        stbuf->st_nlink = 2;
        stbuf->st_ctime = tlx_ctime;
        stbuf->st_mtime = tlx_mtime;
        return 0;
    }

    tlx_reading * p = tlx_get_reading(path + 1);
    if (p) {
        stbuf->st_mode = S_IFREG | 0444;
        stbuf->st_nlink = 1;
        stbuf->st_size = 5;
        stbuf->st_ctime = p->ctime;
        stbuf->st_mtime = p->mtime;
        return 0;
    }

    return -ENOENT;
}

static int _readdir(const char *path, void *buf, fuse_fill_dir_t filler,
                         off_t offset, struct fuse_file_info *fi)
{
    if(strcmp(path, "/") != 0)
        return -ENOENT;

    filler(buf, ".", NULL, 0);
    filler(buf, "..", NULL, 0);
    
    tlx_reading * p = tlx_get_root();
    while (p) {
        char id[6]; // max 5-digit id
        sprintf(id, "%u", p->id);
        filler(buf, id, NULL, 0);
        p = p->next;
    }

    return 0;
}

static int _open(const char *path, struct fuse_file_info *fi)
{
    if ((fi->flags & 3) != O_RDONLY)
        return -EACCES;

    if (tlx_get_reading(path + 1) == NULL)
        return -ENOENT;

    return 0;
}

static int _read(const char *path, char *buf, size_t size, off_t offset,
                      struct fuse_file_info *fi)
{
    tlx_reading * p = tlx_get_reading(path + 1);
    if (p == NULL)
        return -ENOENT;

    if (offset < 5) {
        char * hex = "0123456789ABCDEF";
        char raw[5];
        raw[0] = hex[p->raw >> 12];
        raw[1] = hex[(p->raw >> 8) & 0xf];
        raw[2] = hex[(p->raw >> 4) & 0xf];
        raw[3] = hex[p->raw & 0xf];
        raw[4] = '\n';
        if (offset + size > 5)
            size = 5 - offset;
        memcpy(buf, raw + offset, size);
    } else
        size = 0;

    return size;
}

static void * _init(struct fuse_conn_info * conn)
{
    tlx_init();
    return NULL;
}

static struct fuse_operations _oper = {
    .getattr    = _getattr,
    .readdir    = _readdir,
    .open   = _open,
    .read   = _read,
    .init   = _init,
};

int main(int argc, char *argv[])
{
    return fuse_main(argc, argv, &_oper, NULL);
}



tlx00.c
/*
tlx00 - kernel module for Arexx tlx00 series data logger

Copyright (C) 2011 John Hedges

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

#include <linux/version.h>
#include <linux/module.h>
#include <linux/kernel.h>
#include <linux/usb.h>
#include <linux/workqueue.h>
#include <linux/kobject.h>
#if LINUX_VERSION_CODE >= KERNEL_VERSION(3,0,0)
#include <linux/slab.h>
#endif

#undef PDEBUG
#define PDEBUG(fmt, args...)
#ifdef DEBUG
#  undef PDEBUG
#  define PDEBUG(fmt, args...) printk( KERN_DEBUG "tlx00: " fmt, ## args)
#endif

#define DRIVER_AUTHOR "John Hedges, john@drystone.co.uk"
#define DRIVER_DESC "Arexx temperature logger driver for linux (c) 2011"

#define TL500_VENDOR_ID 0x0451
#define TL500_PRODUCT_ID 0x3211

static void sndcomplete(struct urb* urb);
static void rcvcomplete(struct urb* urb);

struct tlx00_dev {
    int bulk_in_endpointAddr;
    int bulk_out_endpointAddr;
    int iobufsize;
    unsigned char *iobuf;
    struct urb* urb;
    struct usb_device* usbdev;
    struct workqueue_struct* workqueue;
    struct kset *kset;
};

struct sensor {
    struct kobject kobj;
    int raw;
    long timestamp;
};

static struct tlx00_dev* dev = NULL;

static void poll(struct work_struct *unused)
{
    PDEBUG("poll\n");
    usb_fill_bulk_urb(dev->urb, dev->usbdev,
                     usb_sndbulkpipe(dev->usbdev, dev->bulk_out_endpointAddr),
                     dev->iobuf, dev->iobufsize, sndcomplete, NULL);
    
    dev->iobuf[0] = 0x03;
    usb_submit_urb(dev->urb, GFP_KERNEL);
}

DECLARE_DELAYED_WORK(pollwork, poll);

static ssize_t sensor_show(struct kobject *kobj, struct attribute *attr, char *buf)
{
    struct sensor *sens;

    sens = container_of(kobj, struct sensor, kobj);
    sprintf(buf, "%d\n", sens->raw);
    return strlen(buf);
}

static void sensor_release(struct kobject *kobj)
{
    kfree(container_of(kobj, struct sensor, kobj));
}

static struct attribute sensor_attrib = {
    .name = "raw",
#if LINUX_VERSION_CODE < KERNEL_VERSION(3,0,0)
    .owner = NULL,
#endif
    .mode = S_IRUGO
};

static struct attribute *sensor_attribs[] = {
    &sensor_attrib,
    NULL
};

static struct sysfs_ops sensor_sysfs_ops = {
    .show = sensor_show,
    .store = NULL
};

static struct kobj_type sensor_kobj_type = {
    .release = sensor_release,
    .sysfs_ops = &sensor_sysfs_ops,
    .default_attrs = sensor_attribs
};
    
static struct kobject *find_kobj(const char *name) {
    struct kobject *kobj;

    list_for_each_entry(kobj, &dev->kset->list, entry) {
        if (!strcmp(kobj->name, name))
            return kobj;
    }
    return NULL;
}

static void rcvcomplete(struct urb* urb)
{
    char buf[6];
    struct kobject *kobj;
    struct sensor *sens;

    PDEBUG("rcvcomplete %d\n", urb->actual_length);
    PDEBUG("%2.2hhx %2.2hhx %2.2hhx %2.2hhx %2.2hhx %2.2hhx %2.2hhx %2.2hhx\n",
        dev->iobuf[0], dev->iobuf[1], dev->iobuf[2], dev->iobuf[3],
        dev->iobuf[4], dev->iobuf[5], dev->iobuf[6], dev->iobuf[7]);
    switch (dev->iobuf[1]) {
    case 0x09:
    case 0x0a:
        sprintf(buf, "%d", (dev->iobuf[3]<<8)+dev->iobuf[2]);
        kobj = find_kobj(buf);
        if (kobj)
            sens = container_of(kobj, struct sensor, kobj);
        else {
            sens = kmalloc(sizeof(struct sensor), GFP_KERNEL);
            memset(sens, 0, sizeof(struct sensor));
            kobj = &sens->kobj;
            kobj->kset = dev->kset;
            if (kobject_init_and_add(kobj, &sensor_kobj_type, NULL, buf))
                printk(KERN_ERR "Failed to add kobject");
        }
        sens->raw = (dev->iobuf[4]<<8)+dev->iobuf[5];
        queue_delayed_work(dev->workqueue, &pollwork, 0);
        break;
    default:
        queue_delayed_work(dev->workqueue, &pollwork, HZ);
    };
}

static void sndcomplete(struct urb* urb)
{
    PDEBUG("sndcomplete %d\n", urb->actual_length);

    usb_fill_bulk_urb(dev->urb, dev->usbdev,
                     usb_rcvbulkpipe(dev->usbdev, dev->bulk_in_endpointAddr),
                     dev->iobuf, dev->iobufsize, rcvcomplete, NULL);
    
    usb_submit_urb(dev->urb, GFP_KERNEL);
}

static void cleanup(void)
{
    struct kobject *kobj, *kobj1;

    if (dev) {
        if (dev->workqueue) {
            cancel_delayed_work(&pollwork);
            flush_workqueue(dev->workqueue);
            destroy_workqueue(dev->workqueue);
        }
        if (dev->urb)
            usb_free_urb(dev->urb);
        if (dev->iobuf)
            kfree(dev->iobuf);
        if (dev->kset) {
            list_for_each_entry_safe(kobj, kobj1, &dev->kset->list, entry) {
                kobject_put(kobj);
            }
            kset_unregister(dev->kset);
        }
        kfree(dev);
        dev = NULL;
    }
}

static int tlx00_probe(struct usb_interface *intf, const struct usb_device_id *id)
{
    struct usb_host_interface * iface_desc;
    struct usb_endpoint_descriptor * endpoint;
    int i, buffer_size;
    int errno;

    PDEBUG("probed\n");
 
    dev = kmalloc(sizeof(struct tlx00_dev), GFP_KERNEL);
    if (!dev) {
        errno = -ENOMEM;
        goto error;
    }
    memset(dev, 0, sizeof(struct tlx00_dev));

    dev->kset = kset_create_and_add("tlx00", NULL, NULL);
    if (!dev->kset) {
        errno = -ENOMEM;
        goto error;
    }

    /* set up the endpoint information */
    /* use only the first bulk-in and bulk-out endpoints */
    iface_desc = intf->cur_altsetting;
    for (i = 0; i < iface_desc->desc.bNumEndpoints; ++i) {
        endpoint = &iface_desc->endpoint[i].desc;

        if (!dev->bulk_in_endpointAddr &&
            (endpoint->bEndpointAddress & USB_DIR_IN) &&
            ((endpoint->bmAttributes & USB_ENDPOINT_XFERTYPE_MASK)
                    == USB_ENDPOINT_XFER_BULK)) {
            /* we found a bulk in endpoint */
            buffer_size = endpoint->wMaxPacketSize;
            dev->iobufsize = buffer_size;
            dev->bulk_in_endpointAddr = endpoint->bEndpointAddress;
            dev->iobuf = kmalloc(buffer_size, GFP_KERNEL);
            if (!dev->iobuf) {
                err("Could not allocate iobuf");
                errno = -ENOMEM;
                goto error;
            }
        }

        if (!dev->bulk_out_endpointAddr &&
            !(endpoint->bEndpointAddress & USB_DIR_IN) &&
            ((endpoint->bmAttributes & USB_ENDPOINT_XFERTYPE_MASK)
                    == USB_ENDPOINT_XFER_BULK)) {
            /* we found a bulk out endpoint */
            dev->bulk_out_endpointAddr = endpoint->bEndpointAddress;
        }
    }
    if (!(dev->bulk_in_endpointAddr && dev->bulk_out_endpointAddr)) {
        err("Could not find both bulk-in and bulk-out endpoints");
        errno = -EINVAL;
        goto error;
    }

    dev->usbdev = interface_to_usbdev(intf);

    dev->urb = usb_alloc_urb(0, GFP_KERNEL);
    if (!dev->urb)
    {
        errno = -ENOMEM;
        goto error;
    }

    dev->iobuf[0] = 0x04;
    usb_bulk_msg(dev->usbdev,
                 usb_sndbulkpipe(dev->usbdev, dev->bulk_out_endpointAddr),
                 dev->iobuf, dev->iobufsize, NULL, 0);

    dev->workqueue = create_singlethread_workqueue("tlx00");
    queue_delayed_work(dev->workqueue, &pollwork, 0);

    return 0;

error:
    cleanup();
    return errno;
}

static void tlx00_disconnect(struct usb_interface *intf)
{
    PDEBUG("disconnect\n");
    cleanup();
}

static struct usb_device_id tlx00_table [  ] = {
    { USB_DEVICE(TL500_VENDOR_ID, TL500_PRODUCT_ID) },
    { }
};

MODULE_DEVICE_TABLE (usb, tlx00_table);

static struct usb_driver tlx00_driver = {
    .name = "tlx00",
    .id_table = tlx00_table,
    .probe = tlx00_probe,
    .disconnect = tlx00_disconnect,
};

static int __init tlx00_init(void)
{
    int result;

    result = usb_register(&tlx00_driver);
    if (result)
        err("usb_register failed. Error number %d", result);

    return result;
}

static void __exit tlx00_exit(void)
{
    usb_deregister(&tlx00_driver);
}

MODULE_AUTHOR( DRIVER_AUTHOR );
MODULE_DESCRIPTION( DRIVER_DESC );
MODULE_LICENSE("GPL v2");
module_init(tlx00_init);
module_exit(tlx00_exit);

